File: j-backend/src/services/inventoryService.js
/**
 * Inventory management service.
 * Handles product CRUD, stock management, and inventory operations.
 */
const { v4: uuidv4 } = require('uuid');
const { Op } = require('sequelize');
const { Product, Category } = require('../models');
const { recordStockMovement, getCurrentStock, adjustStock } = require('../utils/ledger');
const { logAction } = require('../utils/audit');

/**
 * Create a new product.
 */
const createProduct = async (shopId, productData, createdBy = null) => {
  const product = await Product.create({
    id: uuidv4(),
    shop_id: shopId,
    sku: productData.sku,
    name: productData.name,
    description: productData.description,
    category_id: productData.category_id,
    price_mrp: productData.price_mrp,
    cost_price: productData.cost_price,
    min_stock_threshold: productData.min_stock_threshold || 0,
    barcode: productData.barcode,
    reorder_qty: productData.reorder_qty || 0,
    lead_time_days: productData.lead_time_days || 0,
    attributes: productData.attributes || {},
    is_active: productData.is_active !== false,
  });
  
  // Set initial stock if provided
  if (productData.initial_stock > 0) {
    await recordStockMovement({
      shopId,
      productId: product.id,
      changeQty: productData.initial_stock,
      reason: 'purchase',
      createdBy,
      referenceType: 'initial_stock',
      metaData: { initial_setup: true },
    });
  }
  
  await logAction({
    action: 'create_product',
    userId: createdBy,
    shopId,
    objectType: 'product',
    objectId: product.id,
    payload: { sku: productData.sku, name: productData.name },
  });
  
  return product;
};

/**
 * Get a single product by ID.
 */
const getProduct = async (productId, shopId) => {
  const product = await Product.findOne({
    where: { id: productId, shop_id: shopId },
    include: [{ model: Category, as: 'category' }],
  });
  
  if (product) {
    product.current_stock = await getCurrentStock(productId, shopId);
  }
  
  return product;
};

/**
 * List products with filtering and pagination.
 */
const listProducts = async (shopId, options = {}) => {
  const {
    skip = 0,
    limit = 100,
    categoryId = null,
    search = null,
    isActive = null,
    lowStockOnly = false,
  } = options;
  
  const where = { shop_id: shopId };
  
  if (categoryId) {
    where.category_id = categoryId;
  }
  
  if (search) {
    where[Op.or] = [
      { name: { [Op.iLike]: `%${search}%` } },
      { sku: { [Op.iLike]: `%${search}%` } },
      { barcode: { [Op.iLike]: `%${search}%` } },
    ];
  }
  
  if (isActive !== null) {
    where.is_active = isActive;
  }
  
  if (lowStockOnly) {
    where.current_stock = { [Op.lte]: { [Op.col]: 'min_stock_threshold' } };
  }
  
  const { count, rows } = await Product.findAndCountAll({
    where,
    include: [{ model: Category, as: 'category' }],
    order: [['created_at', 'DESC']],
    offset: skip,
    limit,
  });
  
  // Update current_stock from ledger
  for (const product of rows) {
    product.current_stock = await getCurrentStock(product.id, shopId);
  }
  
  return { products: rows, total: count };
};

/**
 * Update a product.
 */
const updateProduct = async (productId, shopId, productData, updatedBy = null) => {
  const product = await getProduct(productId, shopId);
  if (!product) return null;
  
  const updateFields = {};
  const allowedFields = [
    'name', 'description', 'category_id', 'price_mrp', 'cost_price',
    'min_stock_threshold', 'barcode', 'reorder_qty', 'lead_time_days',
    'attributes', 'is_active',
  ];
  
  for (const field of allowedFields) {
    if (productData[field] !== undefined) {
      updateFields[field] = productData[field];
    }
  }
  
  await product.update(updateFields);
  
  await logAction({
    action: 'update_product',
    userId: updatedBy,
    shopId,
    objectType: 'product',
    objectId: productId,
    payload: updateFields,
  });
  
  return product;
};

/**
 * Adjust product stock to a specific quantity.
 */
const adjustProductStock = async (productId, shopId, newQuantity, reason, createdBy = null) => {
  const product = await getProduct(productId, shopId);
  if (!product) return null;
  
  await adjustStock({
    shopId,
    productId,
    newQuantity,
    createdBy,
    reason,
  });
  
  await logAction({
    action: 'adjust_stock',
    userId: createdBy,
    shopId,
    objectType: 'product',
    objectId: productId,
    payload: { new_quantity: newQuantity, reason },
  });
  
  product.current_stock = await getCurrentStock(productId, shopId);
  return product;
};

/**
 * Get inventory summary statistics.
 */
const getInventorySummary = async (shopId) => {
  const products = await Product.findAll({ where: { shop_id: shopId } });
  
  let totalProducts = products.length;
  let itemsInStock = 0;
  let lowStockCount = 0;
  let outOfStock = 0;
  let totalStockValue = 0;
  
  for (const product of products) {
    const stock = await getCurrentStock(product.id, shopId);
    
    if (stock > 0) {
      itemsInStock++;
      totalStockValue += stock * parseFloat(product.cost_price);
    }
    
    if (stock <= product.min_stock_threshold && stock > 0) {
      lowStockCount++;
    }
    
    if (stock <= 0) {
      outOfStock++;
    }
  }
  
  // New products this month
  const monthStart = new Date();
  monthStart.setDate(1);
  monthStart.setHours(0, 0, 0, 0);
  
  const newThisMonth = await Product.count({
    where: {
      shop_id: shopId,
      created_at: { [Op.gte]: monthStart },
    },
  });
  
  return {
    total_products: totalProducts,
    items_in_stock: itemsInStock,
    low_stock_count: lowStockCount,
    out_of_stock: outOfStock,
    total_stock_value: totalStockValue,
    new_this_month: newThisMonth,
  };
};

module.exports = {
  createProduct,
  getProduct,
  listProducts,
  updateProduct,
  adjustProductStock,
  getInventorySummary,
};

File: backend/app/services/inventory_service.py
"""
Inventory management service.
Handles product CRUD, stock management, and inventory operations.
"""
from typing import Optional, List
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func, desc
from uuid import UUID
from decimal import Decimal

from app.models.product import Product
from app.models.categories import Category
from app.models.stock_movement import StockMovement, stock_movement_reason
from app.utils.ledger import record_stock_movement, get_current_stock
from app.utils.audit import log_action
from app.schemas.product import ProductCreate, ProductUpdate


def create_product(
    db: Session,
    shop_id: UUID,
    product_data: ProductCreate,
    created_by: Optional[UUID] = None
) -> Product:
    """
    Create a new product and optionally set initial stock.
    
    Args:
        db: Database session
        shop_id: Shop UUID
        product_data: Product creation data
        created_by: User UUID who created the product
        
    Returns:
        Created Product object
    """
    # Create product
    product = Product(
        shop_id=shop_id,
        sku=product_data.sku,
        name=product_data.name,
        description=product_data.description,
        category_id=product_data.category_id,
        price_mrp=product_data.price_mrp,
        cost_price=product_data.cost_price,
        min_stock_threshold=product_data.min_stock_threshold,
        barcode=product_data.barcode,
        reorder_qty=product_data.reorder_qty,
        lead_time_days=product_data.lead_time_days,
        attributes=product_data.attributes,
        is_active=product_data.is_active,
    )
    
    db.add(product)
    db.flush()  # Get product ID
    
    # Set initial stock if provided
    if product_data.initial_stock > 0:
        record_stock_movement(
            db=db,
            shop_id=shop_id,
            product_id=product.id,
            change_qty=product_data.initial_stock,
            reason=stock_movement_reason.purchase,
            created_by=created_by,
            reference_type="initial_stock",
            metadata={"initial_setup": True}
        )
    
    # Audit log
    log_action(
        db=db,
        action="create_product",
        user_id=created_by,
        shop_id=shop_id,
        object_type="product",
        object_id=product.id,
        payload={"sku": product_data.sku, "name": product_data.name}
    )
    
    db.commit()
    db.refresh(product)
    return product


def get_product(db: Session, product_id: UUID, shop_id: UUID) -> Optional[Product]:
    """
    Get a single product by ID.
    
    Args:
        db: Database session
        product_id: Product UUID
        shop_id: Shop UUID (for security)
        
    Returns:
        Product object or None
    """
    return db.query(Product).filter(
        and_(Product.id == product_id, Product.shop_id == shop_id)
    ).first()


def list_products(
    db: Session,
    shop_id: UUID,
    skip: int = 0,
    limit: int = 100,
    category_id: Optional[UUID] = None,
    search: Optional[str] = None,
    is_active: Optional[bool] = None,
    low_stock_only: bool = False
) -> tuple[List[Product], int]:
    """
    List products with filtering and pagination.
    
    Args:
        db: Database session
        shop_id: Shop UUID
        skip: Number of records to skip
        limit: Maximum number of records to return
        category_id: Filter by category
        search: Search in name, SKU, or barcode
        is_active: Filter by active status
        low_stock_only: Only return products below threshold
        
    Returns:
        Tuple of (products list, total count)
    """
    query = db.query(Product).filter(Product.shop_id == shop_id)
    
    # Apply filters
    if category_id:
        query = query.filter(Product.category_id == category_id)
    
    if search:
        search_pattern = f"%{search}%"
        query = query.filter(
            or_(
                Product.name.ilike(search_pattern),
                Product.sku.ilike(search_pattern),
                Product.barcode.ilike(search_pattern)
            )
        )
    
    if is_active is not None:
        query = query.filter(Product.is_active == is_active)
    
    if low_stock_only:
        query = query.filter(
            Product.current_stock <= Product.min_stock_threshold
        )
    
    # Get total count
    total = query.count()
    
    # Apply pagination and ordering
    products = query.order_by(desc(Product.created_at)).offset(skip).limit(limit).all()
    
    # Update current_stock from ledger for each product
    for product in products:
        product.current_stock = get_current_stock(db, product.id, shop_id)
    
    return products, total


def update_product(
    db: Session,
    product_id: UUID,
    shop_id: UUID,
    product_data: ProductUpdate,
    updated_by: Optional[UUID] = None
) -> Optional[Product]:
    """
    Update a product.
    
    Args:
        db: Database session
        product_id: Product UUID
        shop_id: Shop UUID
        product_data: Update data
        updated_by: User UUID
        
    Returns:
        Updated Product object or None if not found
    """
    product = get_product(db, product_id, shop_id)
    if not product:
        return None
    
    # Update fields
    update_data = product_data.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(product, field, value)
    
    # Audit log
    log_action(
        db=db,
        action="update_product",
        user_id=updated_by,
        shop_id=shop_id,
        object_type="product",
        object_id=product_id,
        payload=update_data
    )
    
    db.commit()
    db.refresh(product)
    return product


def adjust_product_stock(
    db: Session,
    product_id: UUID,
    shop_id: UUID,
    new_quantity: int,
    reason: str,
    created_by: Optional[UUID] = None
) -> Optional[Product]:
    """
    Adjust product stock to a specific quantity.
    
    Args:
        db: Database session
        product_id: Product UUID
        shop_id: Shop UUID
        new_quantity: Target stock quantity
        reason: Reason for adjustment
        created_by: User UUID
        
    Returns:
        Updated Product object or None if not found
    """
    from app.utils.ledger import adjust_stock
    
    product = get_product(db, product_id, shop_id)
    if not product:
        return None
    
    adjust_stock(
        db=db,
        shop_id=shop_id,
        product_id=product_id,
        new_quantity=new_quantity,
        created_by=created_by,
        reason=reason
    )
    
    # Audit log
    log_action(
        db=db,
        action="adjust_stock",
        user_id=created_by,
        shop_id=shop_id,
        object_type="product",
        object_id=product_id,
        payload={"new_quantity": new_quantity, "reason": reason}
    )
    
    db.commit()
    db.refresh(product)
    product.current_stock = get_current_stock(db, product_id, shop_id)
    return product


def get_inventory_summary(db: Session, shop_id: UUID) -> dict:
    """
    Get inventory summary statistics.
    
    Args:
        db: Database session
        shop_id: Shop UUID
        
    Returns:
        Dictionary with inventory statistics
    """
    products = db.query(Product).filter(Product.shop_id == shop_id).all()
    
    total_products = len(products)
    items_in_stock = sum(1 for p in products if get_current_stock(db, p.id, shop_id) > 0)
    low_stock_count = sum(
        1 for p in products
        if get_current_stock(db, p.id, shop_id) <= p.min_stock_threshold
        and get_current_stock(db, p.id, shop_id) > 0
    )
    out_of_stock = sum(1 for p in products if get_current_stock(db, p.id, shop_id) <= 0)
    
    # Calculate total stock value
    total_stock_value = sum(
        Decimal(str(get_current_stock(db, p.id, shop_id))) * p.cost_price
        for p in products
    )
    
    # New products this month
    from datetime import datetime, timedelta
    month_start = datetime.utcnow().replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    new_this_month = db.query(Product).filter(
        and_(
            Product.shop_id == shop_id,
            Product.created_at >= month_start
        )
    ).count()
    
    return {
        "total_products": total_products,
        "items_in_stock": items_in_stock,
        "low_stock_count": low_stock_count,
        "out_of_stock": out_of_stock,
        "total_stock_value": float(total_stock_value),
        "new_this_month": new_this_month,
    }

